textproc/lilypond-devel/Portfile

```diff
@@ -21,59 +22,150 @@ long_description    Lilypond is a unix-based automated engraving system that \
+                your TeX distribution's binary directory to 'binpath'\
+                in 'macports.conf'?"
+        }
+    } elseif {${lilyponddevel.mactex_bin} eq \
+                  [file dirname [file normalize "${prefix}/bin/kpsewhich"]]} {
+        pre-fetch {
+            return -code error "Variant 'mactex' doesn't work with\
+                installed MacPorts TeXLive packages.  Either\
+                uninstall them or don't use the 'mactex' variant of\
+                lilyponddevel."
+        }
+    } else {
+        set lilyponddevel.have_texgyre \
+            [regexp -line {installed: *Yes} \
+                 [exec ${lilyponddevel.mactex_bin}/tlmgr info --only-installed tex-gyre]]
+    }
+
+    if {${lilyponddevel.have_texgyre}} {
+        set lilyponddevel.texgyredir \
+            [file dirname \
+                [exec ${lilyponddevel.mactex_bin}/kpsewhich texgyreschola-regular.otf]]
+    } else {
+        pre-fetch {
+            return -code error "TeXLive package 'tex-gyre' not\
+                installed; aborting.  Please install it, then\
+                try again"
+        }
+    }
+
+    depends_lib-delete port:texlive \
+                       port:texlive-fonts-recommended \
+                       port:texlive-lang-cyrillic \
+                       port:texlive-metapost
+
+    configure.args-append --with-texgyre-dir=${lilyponddevel.texgyredir}
+}
+
+patchfiles          patch-scripts-build-mf2pt1.pl.diff \
+                    patch-no-Wstrict-prototypes-warning.diff
 post-patch {
     reinplace s|__vector|lily_vector|g ${worksrcpath}/flower/include/std-vector.hh
 
     # Use guile18 header files
-    reinplace s|libguile\.h|libguile18.h|g ${worksrcpath}/configure \
+    reinplace s|libguile\.h|libguile18.h|g \
+        ${worksrcpath}/aclocal.m4 \
+        ${worksrcpath}/configure.ac \
         ${worksrcpath}/lily/include/lily-guile.hh
 
     # help fontforge date calculation
     reinplace s|20110222|2011-02-22|g ${worksrcpath}/configure
 }
 
-# https://lists.gnu.org/archive/html/bug-lilypond/2016-07/msg00071.html
+# lilypond currently relies on a g++ extension (which appears to be
+# non-standard C++) that prevents compilation with clang, cf.
+#
+#   https://lists.gnu.org/archive/html/bug-lilypond/2016-07/msg00071.html
+#
+# and the thread starting with
+#
+#   http://lists.llvm.org/pipermail/cfe-users/2018-November/001417.html
+#
+# According to one of the main developers of lilypond, it is not
+# possible to circumvent this issue without rewriting a considerable
+# amount of code.
+
 # blacklist any clang compiler and any old gcc (a better expression is needed)
-compiler.blacklist *clang* *llvm-gcc* *apple-gcc* gcc gcc-3.3 gcc-4.0 gcc-4.2
-compiler.whitelist macports-gcc-6 macports-gcc-5 macports-gcc-4.9 macports-gcc-4.8 macports-gcc-4.7
+compiler.blacklist \
+    *clang* \
+    *llvm-gcc* \
+    *apple-gcc* \
+    gcc \
+    gcc-3.3 gcc-4.0 gcc-4.2
+compiler.whitelist \
+    macports-gcc-8 \
```

Comment:
Does lilypond still compile on 10.5/ppc?
gcc 8 is almost broken there.

Reply:
No idea, sorry.  Do I have to change something here?

Reply:
gcc8 is broken on PPC systems at present, and also on Leopard Intel IIRC. 

gcc6 and gcc7 work great on those systems for now, and we usually use gcc6 as it has a better success rate in the test suite.

Reply:
OK, thanks.  I repeat my question: Do I have to change something?

Reply:
The basic idea is to blacklist compilers that you know don't work, by appending them to the blacklist, and leave the rest up to the MacPorts logic to sort out. Then many things you may not know about other systems are all take care of magically behind the scenes. So you would not really whitelist a compiler unless it's a special case.

Sometimes, MacPort's list of "fallback" compilers is not long enough to come up with one that is not blacklisted. Then someone who knows what works and what doesn't on a given system will add more fallback compilers by appending them to the fallback list.

You have to be careful about broad generalizations that are likely not true, like blacklisting *gcc*. Also, as you can see, wildcards work, to make life a bit simpler. 

There are premade "groups" of blacklisting and fallback patterns in PortGroups -- the only one to know about just now is the cxx11 1.1 PortGroup, which sets up the scene for a modern compiler supporting c++11, and takes care of everything in the background to do that for you properly.

Reply:
So this port already does a lot of that. I think I would have tried `compiler.fallback-append` rather than `compiler.whitelist`, but anyway, should work. And frankly, if 10.5 PPC doesn't build it, no big deal, and it'll most likely be me that fixes it in the end anyway :>